File: src/core/patterns.ts
/**
 * Ported from restim-desktop/qt_ui/patterns/threephase/
 */

export interface PatternPosition {
  x: number;
  y: number;
}

export interface ThreephasePattern {
  name(): string;
  update(dt: number): PatternPosition;
}

export class CirclePattern implements ThreephasePattern {
  private angle: number = 0;
  private amplitude: number;
  private velocity: number;

  constructor(amplitude: number = 1.0, velocity: number = 1.0) {
    this.amplitude = amplitude;
    this.velocity = velocity;
  }

  name(): string {
    return "Circle";
  }

  update(dt: number): PatternPosition {
    this.angle += dt * this.velocity;
    
    // Normalize angle to keep it within [0, 2*PI] to avoid precision issues over long runs
    this.angle = this.angle % (2 * Math.PI);

    const x = Math.cos(this.angle) * this.amplitude;
    const y = Math.sin(this.angle) * this.amplitude;
    
    return { x, y };
  }
  
  setAmplitude(amplitude: number) {
    this.amplitude = amplitude;
  }
  
  setVelocity(velocity: number) {
    this.velocity = velocity;
  }
}

File: restim-maui/Core/Patterns.cs
using System;

namespace RestimMaui.Core
{
    public struct PatternPosition
    {
        public double X;
        public double Y;
    }

    public interface IThreephasePattern
    {
        string Name { get; }
        PatternPosition Update(double dt);
        void SetVelocity(double velocity);
    }

    public class CirclePattern : IThreephasePattern
    {
        private double _angle = 0;
        private double _amplitude;
        private double _velocity;

        public CirclePattern(double amplitude = 1.0, double velocity = 1.0)
        {
            _amplitude = amplitude;
            _velocity = velocity;
        }

        public string Name => "Circle";

        public PatternPosition Update(double dt)
        {
            _angle += dt * _velocity;
            _angle %= (2 * Math.PI);

            var x = Math.Cos(_angle) * _amplitude;
            var y = Math.Sin(_angle) * _amplitude;

            return new PatternPosition { X = x, Y = y };
        }

        public void SetAmplitude(double amplitude)
        {
            _amplitude = amplitude;
        }

        public void SetVelocity(double velocity)
        {
            _velocity = velocity;
        }
    }
}
